### PR TITLE
Check for attribute-variable correspondence

### LINKED ISSUE #2100: We broke the 1:1 correspondence with attribute names and variable dict names
from @levskaya:

```py
class Foo(nn.Module):
  def setup(self):
    self.foo = nn.Dense(1, name="bar")
    self.qup = self.param('baz', lambda k: jnp.zeros((1,)))
  def __call__(self, x):
    return self.foo(x) + self.qup
Foo().init(jax.random.PRNGKey(0), jnp.zeros((1,)))
```

```
FrozenDict({
    params: {
        baz: DeviceArray([0.], dtype=float32),
        bar: {
            kernel: DeviceArray([[-0.58376074]], dtype=float32),
            bias: DeviceArray([0.], dtype=float32),
        },
    },
})
```

The above used to break loudly, and it should!

Initial investigation by @avital:

Over the past 1.5 years (I ran a test against every single commit), we actually never had any commit where the following code raised an exception:

```py
  def test_setattr_name_var_agreement_in_setup(self):
    class Foo(nn.Module):
      def setup(self):
        self.qup = self.param('baz', lambda k: 0)
      def __call__(self):
        pass

    Foo(parent=None).init(jax.random.PRNGKey(0))
```

But we did, in the part, disallow entirely the use of `name=` for submodules defined within `setup`, which would have disallowed setting the wrong name for a submodule in setup. We lost that guard with https://github.com/google/flax/pull/976/files

I don't think we ever had tests for the variable attribute correspondence. We do have tests that you can't define two variables with the same name in different collections but not that the name aligns with the attribute being assigned to. 

Suggestion from @jheek:

I think you could do something like this to disallowed giving different names in setup.

```py
def __setattr__(self, name, value):
  if any(name in variables[col] for in col):
   assert variables[col] is value, f"A variable named {name} already exist. We don't allow variables and fields to have overlapping names"
```

### PR DESCRIPTION
**Pull request by cgarciae**
Fixes #2100, see discussion.

# Changes
This PR tries to prevent users from assigning fields during `setup` to variables created with different names e.g:

```python
class Foo(nn.Module):
  def setup(self):
    # this would error as value from 'baz' was found for field 'bar'
    self.bar = self.param("baz", lambda key: jnp.array(1))
```
However there are edge cases. The following will yield an error but its not clear it should:

```python
class Foo(nn.Module):
  def setup(self):
    self.bar = self.param("bar", lambda key: jnp.array(1))
    self.baz = self.bar # error: value found but 'baz' key was not found
```
The correspondence can also be broken while being undetectable:
```python
class Foo(nn.Module):
  def setup(self):
    self.bar = self.param("bar", lambda key: jnp.array(1)) + 3
```
In this example no error arises because its value was not found in `variables`.

### CODE REVIEW

**Comment by marcvanzee** [comment]
Thanks for the quick fix!

Following @levskaya's comment on the issue, I am going to run internal tests on this PR to see how many internal clients we break.

**Comment by marcvanzee** [comment]
@cgarciae I ran internal tests and there were a few failures but not as many as I expected. I will try to fix those targets and let you know once it is done.

**Comment by marcvanzee** [comment]
Assigning this priority P1, but we aren't sure whether this will go through internally. We are currently discussing this.

File: flax/linen/module.py [hunk 1]
```diff
@@ -396,6 +396,54 @@ def _get_unbound_fn(method_or_fn: Callable[..., Any]) -> Callable[..., Any]:
...
+  )
+
+  for module in modules:
+    if module.name is not None and module.name != name:
+      raise RuntimeError(
+        f'Module name \'{module.name}\' does not match field name \'{name}\''
+      )
+
+  
+  value_found = False
```
Comment (jheek) [inline flax/linen/module.py:425]: I'm not sure how this deals with nested params, e.g.: self.theta = {"kernel": self.param("kernel", ...)}

**Review by jheek — CHANGES_REQUESTED** [review]

File: tests/linen/linen_module_test.py [hunk 2]
```diff
@@ -1517,5 +1517,66 @@ def __call__(self):
...
+
+    # check variables
+    self.assertIn("bar", variables["params"])
+    self.assertIn("submod_0", variables["params"])
+
+  def test_correspondance_allow_repeated_values_from_fields(self):
+    class Foo(nn.Module):
+      def setup(self):
+        self.bar = self.param("bar", lambda key: jnp.array(1))
+        # this will fail because assigned value is found but
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply tests/linen/linen_module_test.py:1569]: I have to update this comment, @marcvanzee recently requested for this edge case to pass.

File: flax/linen/module.py [hunk 3]
```diff
@@ -396,6 +396,54 @@ def _get_unbound_fn(method_or_fn: Callable[..., Any]) -> Callable[..., Any]:
...
+
+  """
+  # check for Module names
+  modules = (
+    m for m in jax.tree_leaves(val, is_leaf=lambda x: isinstance(x, Module)) 
+    if isinstance(m, Module)
+  )
+
+  for module in modules:
+    if module.name is not None and module.name != name:
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/module.py:419]: My understanding is that this is true in compact whenre you get e.g. `Dense_0`, `Dense_1`, ..., but this shouldn't happen inside `setup`. Please correct me if I am wrong.

File: flax/linen/module.py (hunk 1 above)
Comment (cgarciae) [reply flax/linen/module.py:425]: #2100 is trying to solve very simple breaks in correspondance:

```python
class Foo(nn.Module):
  def setup(self):
    self.foo = nn.Dense(1, name="bar")
    self.qup = self.param('baz', lambda k: jnp.zeros((1,)))
```
Your example involves additional operations over the value which I would consider valid. Base case for this would be:

```python
self.bar = self.param("bar", lambda key: jnp.array(1))
self.baz = self.bar + 1 # valid 
self.foo = {'foo': self.baz} # valid
self.fii = self.foo # recently made valid
```

**Comment by cgarciae** [comment]
Closing this as we are actually going in the opposite direction.
